addons/mass_mailing/models/mass_mailing.py

```diff
@@ -387,6 +388,25 @@ def _compute_statistics(self):
             row['bounced_ratio'] = 100.0 * row['bounced'] / total
             self.browse(row.pop('mailing_id')).update(row)
 
+    @api.multi
+    def _unsubscribe_token(self, res_id, email):
+        """Generate a secure hash for this mailing list and parameters.
+
+        This is appended to the unsubscription URL and then checked at
+        unsubscription time to ensure no malicious unsubscriptions are
+        performed.
+
+        :param int res_id:
+            ID of the resource that will be unsubscribed.
+
+        :param str email:
+            Email of the resource that will be unsubscribed.
+        """
+        icp = self.env["ir.config_parameter"].sudo()
+        salt = icp.get_param("database.secret")
+        source = (self.env.cr.dbname, self.id, res_id, email, salt)
+        return sha256(",".join(map(unicode, source))).hexdigest()
```

Comment:
Rolling out hand-crafted message signatures is generally considered a bad practice. Would you mind using a known secure HMAC implementation, like we're doing in other places? We sometimes add a signed parameter with an expiration date, to implement token expiration, but that seems overkill here. You can find an example of HMAC use [here](https://github.com/odoo/odoo/blob/46be8b30d39de9abaeaa282458d17008d52ccf1d/addons/mail/models/mail_thread.py#L577).